src/coreclr/vm/comdelegate.cpp

```diff
@@ -1941,7 +1908,55 @@ PCODE COMDelegate::GetInvokeMethodStub(EEImplMethodDesc* pMD)
         if (*pMD->GetSig() != (IMAGE_CEE_CS_CALLCONV_HASTHIS | IMAGE_CEE_CS_CALLCONV_DEFAULT))
             COMPlusThrow(kInvalidProgramException);
 
-        ret = COMDelegate::TheDelegateInvokeStub();
+        Stub *pStub = pClass->m_pSingleCastInvokeStub;
+        if (pStub == NULL)
+        {
+            MetaSig sig(pMD);
+
+            BOOL fReturnVal = !sig.IsReturnTypeVoid();
+
+            SigTypeContext emptyContext;
+            ILStubLinker sl(pMD->GetModule(), pMD->GetSignature(), &emptyContext, pMD, (ILStubLinkerFlags)(ILSTUB_LINKER_FLAG_STUB_HAS_THIS | ILSTUB_LINKER_FLAG_TARGET_HAS_THIS));
+
+            ILCodeStream *pCode = sl.NewCodeStream(ILStubLinker::kDispatch);
+
+            // This stub is only used for rare indirect cases, for example
+            // when Delegate.Invoke method is wrapped into another delegate.
+            // Direct invocation of delegate is expanded by JIT.
+            // Emit a recursive call here to let JIT handle complex cases like
+            // virtual dispatch and GC safety.
+
+            // Load the delegate object
+            pCode->EmitLoadThis();
+
+            // Load the arguments
+            for (UINT paramCount = 0; paramCount < sig.NumFixedArgs(); paramCount++)
+                pCode->EmitLDARG(paramCount);
+
+            // recursively call the delegate itself
+            pCode->EmitCALL(pCode->GetToken(pMD), sig.NumFixedArgs(), fReturnVal);
+
+            // return
+            pCode->EmitRET();
+
+            PCCOR_SIGNATURE pSig;
+            DWORD cbSig;
+            pMD->GetSig(&pSig,&cbSig);
+
+            MethodDesc* pStubMD = ILStubCache::CreateAndLinkNewILStubMethodDesc(pMD->GetLoaderAllocator(),
+                                                                   pMD->GetMethodTable(),
+                                                                   ILSTUB_SINGLECASTDELEGATE_INVOKE,
+                                                                   pMD->GetModule(),
+                                                                   pSig, cbSig,
+                                                                   NULL,
+                                                                   &sl);
+
+            pStub = Stub::NewStub(JitILStub(pStubMD));
+
+            InterlockedCompareExchangeT<PTR_Stub>(&pClass->m_pSingleCastInvokeStub, pStub, NULL);
+        }
+
+        ret = pStub->GetEntryPoint();
```

Comment:
I see #105584 is selecting the winning stub from compare exchange. Is it necessary?
Can there be memory leak for the losing one?

And also, is GC transition needed here?

Reply:
> I see https://github.com/dotnet/runtime/pull/105584 is selecting the winning stub from compare exchange. Is it necessary?

I can get confusing e.g. for debugger when there are multiple bodies for a single stub.

> Can there be memory leak for the losing one?

Yes, the losing one will leak. We use similar pattern in other places. It is small limited leak.